plugins/package-managers/pub/src/main/kotlin/Pub.kt

```diff
@@ -132,7 +132,15 @@ class Pub(
 
     private val flutterAbsolutePath = flutterHome.resolve("bin")
 
-    private val gradleFactory = analyzerConfig.determineEnabledPackageManagers().find { it.type.startsWith("Gradle") }
+    private val gradleFactory = analyzerConfig.determineEnabledPackageManagers()
+        .filter { it.type.startsWith("Gradle") }
+        .let { managers ->
+            require(managers.size < 2) {
+                "All of the $managers managers are able to manage 'Gradle' projects. Please enable only one of them."
+            }
+
+            managers.firstOrNull()
+        }
```

Comment:
> If the check fails, this code [1](https://github.com/oss-review-toolkit/ort/blob/8b030ab9513e410e2d79d5883d227b16cd3215e4/analyzer/src/main/kotlin/Analyzer.kt#L100-L104) will throw.

I cannot follow. Why would the linked code to "Associate mapped files by the package manager that manages them" throw in this case? What *would* throw is the more generic check done [here](https://github.com/oss-review-toolkit/ort/blob/8b030ab9513e410e2d79d5883d227b16cd3215e4/model/src/main/kotlin/DependencyGraphNavigator.kt#L52-L55).

> avoid code which handles interdependencies between enabled managers?

I don't see how this could be done. In general I believe that 1) we need to support dependencies between package managers, 2) it makes sense to deal with those dependencies in a generic way / central place as much as possible.

> right in the analyzer to fail early because of mis-configuration

I do see value in consolidating this check and the check from [here](https://github.com/oss-review-toolkit/ort/blob/8b030ab9513e410e2d79d5883d227b16cd3215e4/model/src/main/kotlin/DependencyGraphNavigator.kt#L52-L55) at a "higher level" to fail even before the analyzers are run / instantiated. However, that would be a bigger refactoring to make it work with the server, too, which does not use the `AnalyzerCommand`. Essentially, we'd need to refactor the API functions so that they call `determineEnabledPackageManagers()` themselves, and take an `AnalyzerConfiguration` instead of a list of package managers.

This was just supposed to be a small change to align the logic between two existing checks. I'm a little bit surprised about the stir it creates.

Reply:
> I cannot follow. Why would the linked code to "Associate mapped files by the package manager that manages them" throw in this case? What _would_ throw is the more generic check done [here](https://github.com/oss-review-toolkit/ort/blob/8b030ab9513e410e2d79d5883d227b16cd3215e4/model/src/main/kotlin/DependencyGraphNavigator.kt#L52-L55).

The reason it would throw is IIUC:  `factory.create()` calls the introduced `require` check, because that check is executed as part of the construction of the instance of `Pub`.

> I don't see how this could be done. In general I believe that 1) we need to support dependencies between package managers, 2) it makes sense to deal with those dependencies in a generic way / central place as much as possible.

Sorry, I meant exactly the centralization thing. E.g. try to avoid putting code handling interdependencies of managers into a specific manager implementation.

> However, that would be a bigger refactoring to make it work with the server, too, which does not use the AnalyzerCommand.

I'm not into ORT server development, and this was not obvious to me at all.

> I'm a little bit surprised about the stir it creates.

Not sure why this is perceived as so much stir. I only wanted to understand why you choose to put it into pub and not into a central place.

Reply:
> I'm not into ORT server development, and this was not obvious to me at all.

As a general note, think of the ORT Server as a substitute for the CLI frontend to ORT. That is, the server does not make use of anything inside the `*cli/` or `plugins/commands/` directories.